src/rgw/rgw_dedup_table.h

```diff
@@ -0,0 +1,137 @@
+#pragma once
+#include <cstdint>
+#include <cstddef>
+#include <iterator>
+#include "common/dout.h"
+#include "rgw_dedup_store.h"
+namespace rgw::dedup {
+
+  // 22 Bytes key
+  struct key_t {
+    key_t() { ;}
+    key_t(uint64_t _md5_high,
+          uint64_t _md5_low,
+          uint32_t _size_4k_units,
+          uint16_t _num_parts) {
+      md5_high      = _md5_high;
+      md5_low       = _md5_low;
+      size_4k_units = _size_4k_units;
+      num_parts     = _num_parts;
+    }
+    bool operator==(const struct key_t& other) const {
+      return (this->md5_high       == other.md5_high &&
+              this->md5_low        == other.md5_low  &&
+              this->size_4k_units  == other.size_4k_units &&
+              this->num_parts      == other.num_parts);
+    }
+
+    bool operator!=(const struct key_t& other) const {
+      return !operator==(other);
+    }
+
+    uint64_t hash() const {
+      // The MD5 is already a hashing function so no need for another hash
+      return this->md5_low;
+    }
+
+    uint64_t md5_high;      // High Bytes of the Object Data MD5
+    uint64_t md5_low;       // Low  Bytes of the Object Data MD5
+    uint32_t size_4k_units; // Object size in 4KB units max out at 16TB (AWS MAX-SIZE is 5TB)
```

Comment:
Do we strictly enforce a 5TB object size limit?  When I looked recently in response to a user question I couldn't find any explicit limit.  We have a minimum MPU chunk size, but I didn't see a max.

Reply:
This is something AWS published -
"Individual Amazon S3 objects can range in size from a minimum of 0 bytes to a maximum of 5 TB..."
https://repost.aws/questions/QUe5bHvhVkTuG5PY_Q-DweIQ/s3-storage-capacity#:~:text=The%20total%20volume%20of%20data,single%20PUT%20is%205%20GB.

Reply:
Aye but RGW doesn't AFAICT enforce that limit.  Mind you it wouldn't be a bad idea to do so, but I don't think we do today.

Reply:
An object bigger than 16TB might create a false match when the MD5 happens to be identical to a smaller object, but that will be filtered out before dedup when we check the byte sizes of SRC and DEST.
Please keep in mind that the chance of such a match is crazy low as MD5 128bits hash should be enough as a predictor for a match even without comparing sizes.
Even without that check it will fail the SHA256 so no real problem here